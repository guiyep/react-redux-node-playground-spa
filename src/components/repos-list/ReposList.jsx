import React, { memo, useEffect, useMemo, Fragment } from 'react';
import PropTypes from 'prop-types';
import {
  ReposList_container,
  ReposList_row,
  ReposList_cell,
  ReposList_header,
  ReposList_group_by,
} from './ReposList.module.css';

const buildHeader = columns => (
  <div className={ReposList_header}>
    {columns.map((columnName, index) => (
      <div key={index} className={ReposList_cell}>
        {`${columnName.charAt(0).toUpperCase()}${columnName.slice(1)}`}
      </div>
    ))}
  </div>
);

const buildRows = (columns, list) => (
  <Fragment>
    {(list &&
      list.length > 0 &&
      list.map((item, rowIndex) => (
        <div className={ReposList_row} key={rowIndex}>
          {columns.map((columnName, columnIndex) => (
            <div key={columnIndex} className={ReposList_cell}>
              {item[columnName]}
            </div>
          ))}
        </div>
      ))) || <div className={ReposList_row}>No Items to display</div>}
  </Fragment>
);

const ReposList = memo(({ list, onLoad, groupBy, columns }) => {
  useEffect(() => {
    onLoad();
  }, [onLoad]);

  const currentList = useMemo(() => {
    if (groupBy) {
      return Object.values(
        list.reduce((acc, item) => {
          const groupByValue = item[groupBy] || 'Not defined';

          if (!acc[groupByValue]) {
            acc[groupByValue] = {
              groupBy: item[groupBy],
              list: list.filter(listItem => item[groupBy] === listItem[groupBy]),
            };
          }
          return acc;
        }, {}),
      );
    }
    return list;
  }, [list, groupBy]);

  // break component soon
  if (columns && columns.length === 0) {
    return null;
  }

  return (
    <div className={ReposList_container}>
      {!!groupBy && (
        <Fragment>
          {(currentList &&
            currentList.length > 0 &&
            currentList.map((currentItem, index) => (
              <Fragment key={index}>
                <div className={ReposList_group_by}>{currentItem.groupBy}</div>
                {buildHeader(columns)}
                {buildRows(columns, currentItem.list)}
              </Fragment>
            ))) || <div className={ReposList_row}>No Items to display</div>}
        </Fragment>
      )}
      {!groupBy && (
        <Fragment>
          {buildHeader(columns)}
          {buildRows(columns, currentList)}
        </Fragment>
      )}
    </div>
  );
});

ReposList.displayName = 'ReposList';

ReposList.propTypes = {
  list: PropTypes.arrayOf(PropTypes.object),
  onLoad: PropTypes.func,
  groupBy: PropTypes.string,
  columns: PropTypes.PropTypes.arrayOf(PropTypes.string).isRequired,
};

ReposList.defaultProps = {
  list: [],
  onLoad: () => {},
  groupBy: undefined,
};

export default ReposList;
