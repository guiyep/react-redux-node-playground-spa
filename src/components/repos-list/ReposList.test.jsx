import React from 'react';
import { create } from 'react-test-renderer';
import ReposList from './ReposList';
import ReactDOM from 'react-dom';

const list = [
  {
    name: 'react',
    owner: 'afishlee0',
    languages: ['Scala', 'C#'],
  },
  {
    name: 'webpack',
    owner: 'kwenger5',
    languages: ['Scala'],
  },
  {
    name: 'angular',
    owner: 'shonnan1',
    languages: ['Scala'],
  },
];

describe('ReposList', () => {
  test('it matches the snapshot 1', () => {
    const component = create(<ReposList list={list} columns={['name', 'owner', 'languages']} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('it matches the snapshot 2', () => {
    const component = create(<ReposList list={list} columns={['name', 'languages']} groupBy="owner"/>);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('ReposList renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ReposList list={list} columns={['name', 'owner', 'languages']} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  
  it('ReposList with group renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ReposList list={list} columns={['name', 'languages']} groupBy="owner"/>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

});

