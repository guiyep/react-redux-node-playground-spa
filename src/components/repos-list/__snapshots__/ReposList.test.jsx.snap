// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReposList it matches the snapshot 1 1`] = `
<div
  className="ReposList_container"
>
  <div
    className="ReposList_header"
  >
    <div
      className="ReposList_cell"
    >
      Name
    </div>
    <div
      className="ReposList_cell"
    >
      Owner
    </div>
    <div
      className="ReposList_cell"
    >
      Languages
    </div>
  </div>
  <div
    className="ReposList_row"
  >
    <div
      className="ReposList_cell"
    >
      react
    </div>
    <div
      className="ReposList_cell"
    >
      afishlee0
    </div>
    <div
      className="ReposList_cell"
    >
      Scala
      C#
    </div>
  </div>
  <div
    className="ReposList_row"
  >
    <div
      className="ReposList_cell"
    >
      webpack
    </div>
    <div
      className="ReposList_cell"
    >
      kwenger5
    </div>
    <div
      className="ReposList_cell"
    >
      Scala
    </div>
  </div>
  <div
    className="ReposList_row"
  >
    <div
      className="ReposList_cell"
    >
      angular
    </div>
    <div
      className="ReposList_cell"
    >
      shonnan1
    </div>
    <div
      className="ReposList_cell"
    >
      Scala
    </div>
  </div>
</div>
`;

exports[`ReposList it matches the snapshot 2 1`] = `
<div
  className="ReposList_container"
>
  <div
    className="ReposList_group_by"
  >
    afishlee0
  </div>
  <div
    className="ReposList_header"
  >
    <div
      className="ReposList_cell"
    >
      Name
    </div>
    <div
      className="ReposList_cell"
    >
      Languages
    </div>
  </div>
  <div
    className="ReposList_row"
  >
    <div
      className="ReposList_cell"
    >
      react
    </div>
    <div
      className="ReposList_cell"
    >
      Scala
      C#
    </div>
  </div>
  <div
    className="ReposList_group_by"
  >
    kwenger5
  </div>
  <div
    className="ReposList_header"
  >
    <div
      className="ReposList_cell"
    >
      Name
    </div>
    <div
      className="ReposList_cell"
    >
      Languages
    </div>
  </div>
  <div
    className="ReposList_row"
  >
    <div
      className="ReposList_cell"
    >
      webpack
    </div>
    <div
      className="ReposList_cell"
    >
      Scala
    </div>
  </div>
  <div
    className="ReposList_group_by"
  >
    shonnan1
  </div>
  <div
    className="ReposList_header"
  >
    <div
      className="ReposList_cell"
    >
      Name
    </div>
    <div
      className="ReposList_cell"
    >
      Languages
    </div>
  </div>
  <div
    className="ReposList_row"
  >
    <div
      className="ReposList_cell"
    >
      angular
    </div>
    <div
      className="ReposList_cell"
    >
      Scala
    </div>
  </div>
</div>
`;
